dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,607 @@
-	// Claims are listed in order of priority.
+	// FUTURE EXTENSION:
+	//
+	// OneOf contains a list of requests, only one of which must be satisfied.
+	// Requests are listed in order of priority.
 	//
 	// +optional
 	// +listType=atomic
-	OneOf []DeviceClaimDetail `json:"oneOf,omitempty"`
+	// OneOf []ResourceRequestDetail `json:"oneOf,omitempty"` // candidate for a separate KEP in 1.32, not required for 1.31
 }
 
-// DeviceClaimDetail contains the details of how to fulfill a specific
-// request for devices.
-type DeviceClaimDetail struct {
-	// Class is the name of the DeviceClass to which the requested
-	// devices must belong.
+type ResourceRequestDetail struct {
+	*RequestClassReference `json:",inline,omitempty" protobuf:"bytes,1,opt,name=classReference"`
+
+	// Config defines configuration parameters that apply to the requested resource(s).
+	// They are ignored while allocating the claim.
 	//
-	// +required
-	Class string `json:"class"`
+	// +optional
+	// +listType=atomic
+	Config []ConfigurationParameters `json:"config,omitempty" protobuf:"bytes,1,opt,name=config"`
 
 	// AdminAccess indicates that this is a claim for administrative access
-	// to the devices. Claims with AdminAccess are expected to be used for
+	// to the device(s). Claims with AdminAccess are expected to be used for
 	// monitoring or other management services for a device.  They ignore
 	// all ordinary claims to the device with respect to access modes and
-	// any resource allocations. Ability to create these claims is
-	// controlled via ResourceQuota.
+	// any resource allocations. Ability to request this kind of access is
+	// controlled via ResourceQuota in the resource.k8s.io API.
 	//
-	// Default is false. If true, a DeviceClass must be specified so that
-	// the Driver is known, and only the Constraints fields will be taken
-	// into account for device selection. All devices meeting the
-	// Constraints expressions will be made available as part of the claim
-	// and may be assigned to a container.
+	// Can be combined with a range to ask for access to all devices
+	// on a node which match the requrirements.
 	//
-	// NOTE: This cannot appear in class because the quota code does not do
-	// an indirection. Still searching for a better way to handle this use
-	// case, without creating a new top-level type for it.
+	// Default is false.
 	//
 	// +optional
 	AdminAccess *bool `json:"adminAccess,omitempty"`
 
-	// Constraints is a CEL expression that operates on device attributes.
-	// In order for a device to be considered, this CEL expression and the
-	// Constraints expression from the DeviceClass must both be true.
+	// All these match criteria must be satisfied by all devices allocated for this request.
 	//
 	// +optional
-	Constraints *string `json:"constraints,omitempty"`
-
-	// Device claims may be satisfied by choosing multiple devices instead
-	// of just a single device.
+	// +listType=atomic
+	Match []MatchModel `json:"match,omitempty"`
```

Comment:
We will need to really beef up these docs to explain when to use each layer of constraints and match criteria.

Reply:
Changed, see https://github.com/kubernetes-sigs/wg-device-management/pull/14/files#r1622096742.